vault/resource_database_secret_backend_static_role.go

```diff
@@ -0,0 +1,210 @@
+					value := v.(int)
+					if value < 5 {
+						errs = append(errs, fmt.Errorf("The minimum value of rotation_period is 5 seconds."))
+					}
+					return
+				},
+			},
+			"db_name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Database connection to use for this role.",
+			},
+			"rotation_statements": {
+				Type:        schema.TypeList,
+				Required:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "Database statements to execute to rotate the password for the configured database user.",
+			},
+		},
+	}
+}
+
+func databaseSecretBackendStaticRoleWrite(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	backend := d.Get("backend").(string)
+	name := d.Get("name").(string)
+
+	path := databaseSecretBackendStaticRolePath(backend, name)
+
+	data := map[string]interface{}{
+		"username":            d.Get("username"),
+		"rotation_period":     d.Get("rotation_period"),
+		"db_name":             d.Get("db_name"),
+		"rotation_statements": d.Get("rotation_statements"),
+	}
+
+	log.Printf("[DEBUG] Creating static role %q on database backend %q", name, backend)
+	_, err := client.Logical().Write(path, data)
+	if err != nil {
+		return fmt.Errorf("error creating static role %q for backend %q: %s", name, backend, err)
+	}
+	log.Printf("[DEBUG] Created static role %q on AWS backend %q", name, backend)
+
+	d.SetId(path)
+	return databaseSecretBackendStaticRoleRead(d, meta)
+}
+
+func databaseSecretBackendStaticRoleRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	path := d.Id()
+
+	name, err := databaseSecretBackendStaticRoleNameFromPath(path)
+	if err != nil {
+		log.Printf("[WARN] Removing database static role %q because its ID is invalid", path)
+		d.SetId("")
+		return fmt.Errorf("invalid static role ID %q: %s", path, err)
+	}
+
+	backend, err := databaseSecretBackendStaticRoleBackendFromPath(path)
+	if err != nil {
+		log.Printf("[WARN] Removing database static role %q because its ID is invalid", path)
+		d.SetId("")
+		return fmt.Errorf("invalid static role ID %q: %s", path, err)
+	}
+
+	log.Printf("[DEBUG] Reading static role from %q", path)
+	role, err := client.Logical().Read(path)
+	if err != nil {
+		return fmt.Errorf("error reading static role %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Read static role from %q", path)
+	if role == nil {
+		log.Printf("[WARN] Static role %q not found, removing from state", path)
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("backend", backend)
+	d.Set("name", name)
+	d.Set("username", role.Data["username"])
+	d.Set("db_name", role.Data["db_name"])
+
+	if v, ok := role.Data["rotation_period"]; ok {
+		n, err := v.(json.Number).Int64()
+		if err != nil {
+			return fmt.Errorf("unexpected value %q for rotation_period of %q", v, path)
+		}
+		d.Set("rotation_period", n)
```

Comment:
For primitive values, we don't need to worry about the potential err returned from `d.Set`, but I recently learned from one of my coworkers that it would be wise to check it when setting more complex values.

Reply:
Whoops! I meant this for line 155 below where it reads `d.Set("rotation_statements", rotation)`. :blush: